File: MyApplication24/app/src/main/java/com/example/myapplication24/MainAdapter.java
package com.example.myapplication24;

import android.content.Context;
import android.os.Message;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

public class MainAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private List<MessageBean> items = new ArrayList<>();
    private boolean isLoaderVisible = false;
    private static int VIEW_TYPE_MESSAGE = 0;
    private static int VIEW_TYPE_LOADING = 1;

    @Override
    public int getItemCount() {
        Log.e("TAG", "count:" + items.size());
        int count;
        if (items == null) {
            count = 0;
        } else {
            count = items.size();
        }
        return count;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == VIEW_TYPE_MESSAGE) {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_message, parent, false);
            return new MyViewHolder(view);
        } else {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_loading, parent, false);
            return new LoadingViewHolder(view);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (holder.getItemViewType() == VIEW_TYPE_MESSAGE) {
            ((MyViewHolder) holder).description.setText(items.get(position).getDescription());
            ((MyViewHolder) holder).title.setText(items.get(position).getTitle());
        } else {
            //loading点击事件
        }
    }

    @Override
    public int getItemViewType(int position) {
        int result;
        if (isLoaderVisible) {
            if (position == items.size() - 1) {
                result = VIEW_TYPE_LOADING;
            } else {
                result = VIEW_TYPE_MESSAGE;
            }
        } else {
            result = VIEW_TYPE_MESSAGE;
        }
        return result;
    }


    private void add(MessageBean messageBean) {
        items.add(messageBean);
        notifyItemInserted(items.size() - 1);
    }

    public void addAll(List<MessageBean> mlist) {
        //  this.items.addAll(mlist); for循环遍历，遍历list元素内容
        for (MessageBean amessages : mlist) {
            add(amessages);
        }
    }

    public void remove(MessageBean messageBean) {
        int position = items.indexOf(messageBean);
        if (position > -1) {
            items.remove(position);
            notifyItemRemoved(position);
        }
    }

    public void addLoading() {
        isLoaderVisible = true;
        add(new MessageBean("张飞", "赵云"));
    }

    public void removeLoading() {
        isLoaderVisible = false;
        int position = items.size() - 1;
        MessageBean item = getItem(position);

        if (item != null) {
            items.remove(position);
            notifyItemRemoved(position);
        }
    }

    private MessageBean getItem(int position) {
        return items.get(position);
    }

}

File: MyApplication23/app/src/main/java/com/example/myapplication23/MainAdapter.kt
package com.example.myapplication23


import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView

class MainAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private var items: MutableList<Message> = mutableListOf()
    private var isLoaderVisible = false

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val layoutInflater = LayoutInflater.from(parent.context)

        return when (viewType) {
            VIEW_TYPE_MESSAGE -> {
                val view = layoutInflater.inflate(R.layout.item_message, parent, false)
                ItemViewHolder(view)
            }
            else -> {
                val view = layoutInflater.inflate(R.layout.item_loading, parent, false)
                LoadingViewHolder(view)
            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (holder.itemViewType) {
            VIEW_TYPE_MESSAGE -> {
                val item = holder as ItemViewHolder
                item.bind(items[position])
            }
            VIEW_TYPE_LOADING -> {
                val item = holder as LoadingViewHolder
                item.bind()
            }
        }
    }

    override fun getItemCount(): Int {
        return when(items) {
            null -> 0
            else -> items.size
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (isLoaderVisible) {
            when (position) {
                items.size -1 -> VIEW_TYPE_LOADING
                else -> VIEW_TYPE_MESSAGE
            }
        } else {
            VIEW_TYPE_MESSAGE
        }
    }

    inner class ItemViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val title = itemView.findViewById<TextView>(R.id.tv_title)
        private val description = itemView.findViewById<TextView>(R.id.tv_description)

        fun bind(item: Message) {
            title.text = item.title
            description.text = item.description
        }
    }

    inner class LoadingViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val progressBar = itemView.findViewById<ProgressBar>(R.id.progress_bar)

        fun bind() {}
    }

    private fun add(message: Message) {
        items.add(message)
        notifyItemInserted(items.size -1)
    }

    fun addAll(messages: List<Message>) {
        for (items in messages) {
            add(items)
        }
    }

    private fun remove(message: Message) {
        val position = items.indexOf(message)

        if (position > -1) {
            items.removeAt(position)
            notifyItemRemoved(position)
        }
    }

    fun addLoading() {
        isLoaderVisible = true
        add(Message("张飞", "赵云"))
    }

    fun removeLoading() {
        isLoaderVisible = false
        val position = items.size - 1
        val item = getItem(position)

        if (item != null) {
            items.removeAt(position)
            notifyItemRemoved(position)
        }
    }

    fun clear() {
        while (itemCount > 0) {
            remove(getItem(0))
        }
    }

    private fun getItem(position: Int): Message {
        return items[position]
    }


    companion object {
        const val VIEW_TYPE_MESSAGE = 0
        const val VIEW_TYPE_LOADING = 1
    }
}